presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java

```diff
@@ -256,23 +260,51 @@ public PrestoSparkQueryExecutionFactory(
         this.nodeMemoryConfig = requireNonNull(nodeMemoryConfig, "nodeMemoryConfig is null");
         this.featuresConfig = requireNonNull(featuresConfig, "featuresConfig is null");
         this.queryManagerConfig = requireNonNull(queryManagerConfig, "queryManagerConfig is null");
+        this.securityConfig = requireNonNull(securityConfig, "securityConfig is null");
         this.waitTimeMetrics = ImmutableSet.copyOf(requireNonNull(waitTimeMetrics, "waitTimeMetrics is null"));
         this.ddlTasks = ImmutableMap.copyOf(requireNonNull(ddlTasks, "ddlTasks is null"));
         this.errorClassifier = requireNonNull(errorClassifier, "errorClassifier is null");
         this.historyBasedPlanStatisticsTracker = requireNonNull(historyBasedPlanStatisticsManager, "historyBasedPlanStatisticsManager is null").getHistoryBasedPlanStatisticsTracker();
     }
 
+    /**
+     * When selectAuthorizedIdentity API is not enabled, we check the delegation permission
+     */
     private void checkPermissions(QueryId queryId, SessionContext sessionContext)
     {
         Identity identity = sessionContext.getIdentity();
-        accessControl.checkCanSetUser(
-                identity,
-                new AccessControlContext(
-                        queryId,
-                        Optional.ofNullable(sessionContext.getClientInfo()),
-                        Optional.ofNullable(sessionContext.getSource())),
-                identity.getPrincipal(),
-                identity.getUser());
+        if (!securityConfig.isAuthorizedIdentitySelectionEnabled()) {
+            accessControl.checkCanSetUser(
+                    identity,
+                    new AccessControlContext(
+                            queryId,
+                            Optional.ofNullable(sessionContext.getClientInfo()),
+                            Optional.ofNullable(sessionContext.getSource())),
+                    identity.getPrincipal(),
+                    identity.getUser());
+        }
+    }
+
+    /**
+     * When selectAuthorizedIdentity API is enabled,
+     * 1. Check the delegation permission, which is inside the API call
+     * 2. Select and return the authorized identity
+     */
+    private Optional<AuthorizedIdentity> getAuthorizedIdentity(QueryId queryId, SessionContext sessionContext)
```

Comment:
Can we share the implementation with what we have in DispatchManager, instead of having it in two different places?

Reply:
Thanks for the feedback. I finally got back to this and refactored the shared code to `AccessControlUtils` class.